### PR TITLE
WIP: Batch write implementation

### PR DESCRIPTION
**Pull request by ShauneStark**

### CODE REVIEW

**Comment by ericmj** [comment]
db-conn has been merged into master and we will have new release out shortly. You can rebase this PR against master. What else is needed to finish this?

File: lib/mongobatch/mongobatch.ex [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+    #     %{index: 101, _id: "iuhjunihj"},
+    #   ]
+    # }
+    # ---Mock response generation code above---
+
+    if mongowriteresult["ok"] == 1 do
+      # Create optimistic base result array
+      resultstream = Stream.cycle([%{ok: 1}])
+      resultbaselist = Enum.take(resultstream, numwrittenitems)
+      resultbaselist = Enum.to_list(resultbaselist)
```
Comment (ghost) [inline lib/mongobatch/mongobatch.ex:78]: Use pipe here

File: lib/mongobatch/mongobatch.ex [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+    end
+  end
+  #---------------------------------------------------------------------------#
+  def getpagebreaks(listofdocuments) do
+    # First, Encode submitted docs
+    encodeddocs = Enum.map(
+      listofdocuments,
+      fn(x) ->
+        BSON.encode(x)
+      end
```
Comment (ghost) [inline lib/mongobatch/mongobatch.ex:142]: Replace lambda with `&BSON.encode/1`

File: lib/mongobatch/mongobatch.ex [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+    )
+
+    # Next, create list of doc sizes
+    docsizes = Enum.map(
+      encodeddocs,
+      fn(x) ->
+        #byte_size(x)
+        IO.iodata_length(x)
```
Comment (ghost) [inline lib/mongobatch/mongobatch.ex:150]: Replace lambda with `&IO.iodata_length/1`

File: lib/mongobatch/mongobatch.ex [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+    # Process writes
+    if writeorder == false do
+      # Sort by write type
+      {insertdoclist, updatedoclist, deletedoclist} = BatchWriteSorter.sortunordered(writedocument.writes)
+
+      # Process inserts first
+      # Build insert document template
+      inserttemplate = %{BatchInsert.defaults() | insert: writecollection}
+      inserttemplate = %{inserttemplate | writeConcern: writeconcern}
+      inserttemplate = %{inserttemplate | ordered: writeorder}
```
Comment (ghost) [inline lib/mongobatch/mongobatch.ex:342]: Combine these

File: lib/mongobatch/mongobatch.ex [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+        [submittomongo(%{inserttemplate | documents: insertdoclist})]
+      else
+        insertpages(inserttemplate, insertdoclist, insertfullpageindices, [])
+      end
+      insertfinalresult = combineresults(insertresultset)
+
+      # Process updates second
+      updatetemplate = %{BatchUpdate.defaults() | update: writecollection}
+      updatetemplate = %{updatetemplate | writeConcern: writeconcern}
+      updatetemplate = %{updatetemplate | ordered: writeorder}
```
Comment (ghost) [inline lib/mongobatch/mongobatch.ex:356]: Combine these

File: lib/mongobatch/mongobatch.ex [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+        [submittomongo(%{updatetemplate | updates: updatedoclist})]
+      else
+        updatepages(updatetemplate, updatedoclist, updatefullpageindices, [])
+      end
+      updatefinalresult = combineresults(updateresultset)
+
+      # Process deletes last
+      deletetemplate = %{BatchDelete.defaults() | delete: writecollection}
+      deletetemplate = %{deletetemplate | writeConcern: writeconcern}
+      deletetemplate = %{deletetemplate | ordered: writeorder}
```
Comment (ghost) [inline lib/mongobatch/mongobatch.ex:370]: Combine these. Don't repeat yourself.

File: lib/mongobatch/mongobatch.ex [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+      %{insertresults: insertfinalresult, updateresults: updatefinalresult, deleteresults: deletefinalresult}
+    else
+      # Parse writedocument into ordered list of contiguous type documents
+      orderedwritegroupmaps = BatchWriteSorter.sortordered(writedocument)
+      # Batch write each writegroup list and return total results
+      Enum.reduce(
+        orderedwritegroupmaps,
+        [],
+        fn(x, resultlist) ->
+          case x.type do
```
Comment (ghost) [inline lib/mongobatch/mongobatch.ex:391]: Don't repeat yourself

File: lib/mongobatch/mongobatch.ex [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+    collection = submitteddocument[documentwritetype]
+
+    # Reorder fields to make sure command is first
+    submitteddocument = submitteddocument
+    |> Map.drop([documentwritetype])
+    |> Enum.into([])
+    |> List.insert_at(0, {documentwritetype, collection})
+
+    # Create connection to mongo
+    {:ok, mpid} = Mongo.start_link(hostname: "mongo1.hellodata.com", database: "test")
```
Comment (ghost) [inline lib/mongobatch/mongobatch.ex:27]: Don't reconnect each submission

File: lib/mongobatch/mongobatch.ex [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+    # if (finalresult.ok == 1) do
+    #   finalresult = Map.delete(finalresult, :error)
+    # end
+    # finalresult
+
+    # Simply concatenate the list of lists into one list
+    Enum.concat(resultset)
+  end
+  #---------------------------------------------------------------------------#
+  def insert(insertdocument) do
```
Comment (ghost) [inline lib/mongobatch/mongobatch.ex:288]: Don't repeat yourself

**Review by ghost — CHANGES_REQUESTED** [review]
Just a quick initial review

**Comment by ericmj** [comment]
Ping?

**Comment by ghost** [comment]
Will be doing more review and refactoring on this shortly, but not until after #93

**Comment by ericmj** [comment]
Cool, didn't realise you were handling this PR @cjhowe.

**Comment by ankhers** [comment]
Is there any update for this now that #124 has been merged?

**Comment by ben-pr-p** [comment]
What's the status of bulk write @ankhers ?

**Comment by ankhers** [comment]
Unfortunately the original author stopped work on this. There has been no work on it since.

**Comment by ben-pr-p** [comment]
Gotcha. Do you have an understanding what is required to finish it up?

**Comment by ankhers** [comment]
Unfortunately I have never really gone through the code here, nor looked at the spec for this.

**Pull request closed without merging**